lib/recursive-methods.rb

```diff
@@ -1,49 +1,106 @@
 # Authoring recursive algorithms. Add comments including time and space complexity for each method.
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n)
+# Space complexity: O(n)
 def factorial(n)
-    raise NotImplementedError, "Method not implemented"
+    # raise NotImplementedError, "Method not implemented"
+    if n < 0
+        raise ArgumentError
+    elsif n <= 1
+        return 1
+    else
+        return n * factorial(n-1)
+    end
 end
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n)
+# Space complexity: O(n^2)
 def reverse(s)
-    raise NotImplementedError, "Method not implemented"
+    # raise NotImplementedError, "Method not implemented"
+    if s.length <= 1
+        return s
+    else
+        return "#{s[-1] + reverse(s[1..-2]) + s[0]}"
+    end
 end
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n)
+# Space complexity: O(n^2)
 def reverse_inplace(s)
```

Comment:
⚠️ this is not done in-place.  You are creating new strings with each recursive call. 

Consider adjusting the method signature as follows

```ruby
def reverse_inplace(s, low = 0, high = s.length - 1)
```

Can you see any ideas on how to reverse things in place?